### PR TITLE
Registers boolean jinja filters as jinja tests

### PR DESCRIPTION
**Pull request by ghost**
Jinja2 has separate notions of [Filters](http://jinja.pocoo.org/docs/templates/#filters) and [Tests](http://jinja.pocoo.org/docs/templates/#tests), but Ansible currently registers all of its Jinja2 extensions as filters, even those that return boolean values.

This commit ensures that the boolean-returning filters are also registered as tests, and provides hooks for user plugins to create their own tests using a filter_plugin.

It also updates the documentation for jinja test functionality.

### CODE REVIEW

**Comment by ghost** [comment]
This is a reworked version of #8217. See also the [ansible-devel discussion](https://groups.google.com/forum/#!topic/ansible-devel/Oa6XSffcHJg).

File: lib/ansible/utils/__init__.py [hunk 1]
```diff
@@ -1166,11 +1166,13 @@ def safe_eval(expr, locals={}, include_exceptions=False):
...
-    filter_list = []
+    filter_list = set()
     for filter in filter_loader.all():
-        filter_list.extend(filter.filters().keys())
+        # Allow user filter_plugins to omit either filters or tests.
+        filter_list.update(getattr(filter, "filters", list)())
+        filter_list.update(getattr(filter, "tests", list)())
 
-    CALL_WHITELIST = C.DEFAULT_CALLABLE_WHITELIST + filter_list
+    CALL_WHITELIST = C.DEFAULT_CALLABLE_WHITELIST + list(filter_list)
```
Comment (ghost) [inline lib/ansible/utils/__init__.py:1175]: Should `CALL_WHITELIST` be a set? It only seems to be used with the `in` operator below. I didn't want to change more than necessary here, but I'm happy to change this to be a set if you think it's a good idea.

**Comment by ghost** [comment]
I can rebase this to make it mergeable again, but is it worth it? Is there any chance of this actually getting accepted any time soon? Is there anything else I can do to help?

**Comment by hkariti** [comment]
+1 on this. being able to use selectattr can be very useful

**Comment by quixoten** [comment]
+1. The mangling of Jinja's "test" and "filter" concepts in Ansible has been confusing. The inability to define new tests leaves a lot of flexibility on the table.

**Comment by mintern** [comment]
For now, I've given up on this, as I've seen no interest from the maintainer in accepting what I thought was a high quality Pull Request.

**Comment by quixoten** [comment]
@mintern, thanks for the update. I took notes from your work and have a merge-able branch working on my box. I will link it here when I create the PR.

**Comment by mintern** [comment]
Oh, that's great! Thanks!

**Comment by ghost** [comment]
There's not much reason for me to keep this open any more. Best of luck, @quixoten, and best wishes, @mpdehaan.

**Comment by amenonsen** [comment]
Is anyone already working on this for v2? Since this has been reopened, I'm not sure if #9645 supersedes this, or what.

Update: OK, @jimi-c has said he's working on #9645, so this PR should probably be closed now since that one is tagged for v2.

**Comment by bcoca** [comment]
fixed via merged #9645

**Pull request closed without merging**